File: core/src/com/alexpi/awesometanks/entities/blocks/Turret.java
package com.alexpi.awesometanks.entities.blocks;

import com.alexpi.awesometanks.entities.DamageListener;
import com.alexpi.awesometanks.utils.Utils;
import com.badlogic.gdx.assets.AssetManager;
import com.badlogic.gdx.graphics.Texture;
import com.badlogic.gdx.graphics.g2d.Batch;
import com.badlogic.gdx.graphics.g2d.Sprite;
import com.badlogic.gdx.math.Vector2;
import com.badlogic.gdx.physics.box2d.Filter;
import com.badlogic.gdx.physics.box2d.PolygonShape;
import com.badlogic.gdx.physics.box2d.World;
import com.alexpi.awesometanks.utils.Constants;
import com.alexpi.awesometanks.weapons.Weapon;
import com.badlogic.gdx.scenes.scene2d.Group;

/**
 * Created by Alex on 18/02/2016.
 */
public class Turret extends Block {
    private Weapon weapon;
    private Vector2 targetPosition;
    private AssetManager manager;
    private Group entityGroup;
    private final static float ROTATION_SPEED = .035f;
    public Turret(DamageListener listener, AssetManager manager, Group entityGroup, World world, Vector2 targetPosition, int posX, int posY, int type, boolean sound) {
        super(manager,"sprites/turret_base.png",  world,new PolygonShape(),500, posX, posY, .8f, true, listener);
        this.manager = manager;
        this.entityGroup = entityGroup;
        Filter filter = new Filter();
        filter.categoryBits = Constants.CAT_ENEMY;
        fixture.setFilterData(filter);
        this.targetPosition = targetPosition;

        weapon = Weapon.getWeaponAt(type, manager, 1, 3, false, sound);
        weapon.setUnlimitedAmmo(true);
        setOrigin(getWidth() / 2, getHeight() / 2);
    }

    @Override
    public void draw(Batch batch, float parentAlpha) {
        super.draw(batch, parentAlpha);
        batch.draw(weapon.sprite, getX(), getY(), getOriginX(), getOriginY(), getWidth(), getHeight(), getScaleX(), getScaleY(), weapon.sprite.getRotation());
    }

    @Override
    public void act(float delta) {
        super.act(delta);
        float distanceFromTarget  = (float) (Utils.fastHypot(targetPosition.x-body.getPosition().x,targetPosition.y-body.getPosition().y));

        if(distanceFromTarget < 4 && isAlive()){
            weapon.setDesiredAngleRotation(targetPosition.x - body.getPosition().x, targetPosition.y - body.getPosition().y);
            weapon.updateAngleRotation(ROTATION_SPEED);
            if(weapon.hasRotated())weapon.shoot(manager,entityGroup,body.getWorld(),body.getPosition());
        }
    }

}

File: core/src/com/alexpi/awesometanks/entities/blocks/Turret.kt
package com.alexpi.awesometanks.entities.blocks

import com.alexpi.awesometanks.entities.ai.TurretAI
import com.alexpi.awesometanks.entities.ai.TurretAICallback
import com.alexpi.awesometanks.entities.components.body.BodyShape
import com.alexpi.awesometanks.entities.components.body.FixtureFilter
import com.alexpi.awesometanks.entities.items.GoldNugget
import com.alexpi.awesometanks.screens.game.stage.GameContext
import com.alexpi.awesometanks.screens.game.stage.GameStage
import com.alexpi.awesometanks.utils.RandomUtils
import com.alexpi.awesometanks.weapons.Weapon
import com.badlogic.gdx.graphics.g2d.Batch
import com.badlogic.gdx.math.Vector2

/**
 * Created by Alex on 18/02/2016.
 */
class Turret(
    gameContext: GameContext,
    pos: Vector2,
    type: Weapon.Type
) : HealthBlock( gameContext,"sprites/turret_base.png", BodyShape.Box(.8f, .8f), pos, getHealthByType(type),true, true, FixtureFilter.TURRET), TurretAICallback {
    private val weapon: Weapon
    private val gameStage: GameStage = gameContext.getStage()
    private val enemyAI = TurretAI(gameContext, bodyComponent.body.position, this)
    private val nuggetValue: Int = getNuggetValue(type)
    private val explosionManager = gameContext.getExplosionManager()

    override fun drawBlock(batch: Batch, parentAlpha: Float) {
        super.drawBlock(batch, parentAlpha)
        weapon.draw(batch,color, x, parentAlpha, originX, originY, width, height, scaleX, scaleY, y)

    }

    override fun act(delta: Float) {
        if (!healthComponent.isFrozen){
            enemyAI.update()
        } else{
            await()
        }
        super.act(delta)
    }

    companion object {
        private const val ROTATION_SPEED = .035f

        fun getHealthByType(type: Weapon.Type): Float{
            val typeMultiplier: Float = when(type){
                Weapon.Type.MINIGUN -> 0f
                Weapon.Type.SHOTGUN -> .2f
                Weapon.Type.RICOCHET -> .3f
                Weapon.Type.FLAMETHROWER -> .6f
                Weapon.Type.ROCKETS -> .6f
                Weapon.Type.CANNON -> .6f
                Weapon.Type.LASERGUN -> 1f
                Weapon.Type.RAILGUN -> 1f

            }
            return 200f + typeMultiplier * 500f
        }

        private fun getNuggetValue(type: Weapon.Type): Int{
            val typeMultiplier: Float = when(type){
                Weapon.Type.MINIGUN -> 0f
                Weapon.Type.SHOTGUN -> .2f
                Weapon.Type.RICOCHET -> .3f
                Weapon.Type.FLAMETHROWER -> .6f
                Weapon.Type.CANNON -> .6f
                Weapon.Type.ROCKETS -> .6f
                Weapon.Type.LASERGUN -> 1f
                else -> 1f

            }
            return 60 + (typeMultiplier*75).toInt()
        }
    }

    override fun remove(): Boolean {
        dropLoot()
        gameStage.checkLevelCompletion()
        return super.remove()
    }

    private fun dropLoot() {
        val count = RandomUtils.getRandomInt(10, 15)
        repeat(count){
            parent.addActor(GoldNugget(bodyComponent.body.position, RandomUtils.getRandomInt(nuggetValue - 5, nuggetValue + 5)))

        }
    }

    init {
        weapon = Weapon.getWeaponAt(type, gameContext,1f, 2, false)
        weapon.unlimitedAmmo = true
        setOrigin(width / 2, height / 2)
    }

    override fun attack(angle: Float) {
        weapon.desiredRotationAngle = angle
        weapon.updateAngleRotation(ROTATION_SPEED)
        weapon.shoot(parent, bodyComponent.body.position)
    }

    override fun await() {
        weapon.await()
    }
}